File: leet-code/56. Merge Intervals/MergeIntervals56_2.java
package com.example;

import java.util.*;

public class MergeIntervals56_3 {
    public int[][] merge(int[][] intervals) {
        Arrays.sort(intervals, Comparator.comparingInt(a -> a[0]));
        List<int[]> list = new ArrayList<>();
        for (int[] i : intervals) {
            if (!list.isEmpty() && list.get(list.size() - 1)[1] >= i[0]) {
                list.get(list.size() - 1)[1] = Math.max(list.get(list.size() - 1)[1], i[1]);
            } else {
                list.add(i);
            }
        }
        return list.toArray(new int[list.size()][]);
    }
}

File: leet-code/56. Merge Intervals/MergeIntervals56_2.kt
package com.example

class MergeIntervals56_2 {
    fun merge(intervals: Array<IntArray>): Array<IntArray> {
        val sortedIntervals = intervals.sortedArrayWith { a, b -> if (a[0] == b[0]) a[1] - b[1] else a[0] - b[0] }
        val list = mutableListOf<IntArray>()
        for (interval in sortedIntervals) {
            if (list.isEmpty() || list.last()[1] < interval[0]) {
                list.add(interval)
                continue
            }
            if (list.last()[1] < interval[1]) {
                list.last()[1] = interval[1]
            }
        }
        return list.toTypedArray()
    }
}
